src/tqec/position.py

```diff
@@ -60,3 +63,59 @@ class Displacement:
 
     x: int
     y: int
+
+
+@dataclass(frozen=True, order=True)
+class Position3D:
+    """A 3D integer position."""
+
+    x: int
+    y: int
+    z: int
+
+    def __post_init__(self) -> None:
+        if any(not isinstance(i, int) for i in astuple(self)):
+            raise TQECException("Position must be an integer.")
+
+    def shift_by(self, dx: int = 0, dy: int = 0, dz: int = 0) -> "Position3D":
+        """Shift the position by the given offset."""
+        return Position3D(self.x + dx, self.y + dy, self.z + dz)
+
+    def is_neighbour(self, other: "Position3D") -> bool:
+        """Check if the other position is near to this position, i.e. Manhattan distance is 1."""
+        return (
+            abs(self.x - other.x) + abs(self.y - other.y) + abs(self.z - other.z) == 1
+        )
+
+    def as_tuple(self) -> tuple[int, int, int]:
+        """Return the position as a tuple."""
+        return astuple(self)
+
+    def __str__(self) -> str:
+        return f"({self.x},{self.y},{self.z})"
+
+
+class Direction3D(Enum):
+    """Axis directions in the 3D spacetime diagram."""
+
+    X = auto()
+    Y = auto()
+    Z = auto()
+
+    @staticmethod
+    def all() -> list["Direction3D"]:
+        """Get all directions."""
+        return [Direction3D.X, Direction3D.Y, Direction3D.Z]
+
+    @staticmethod
+    def from_axis_index(i: int) -> "Direction3D":
+        """Get the direction from the axis index."""
+        return Direction3D.all()[i]
+
+    @property
+    def axis_index(self) -> int:
+        """Get the axis index."""
+        return Direction3D.all().index(self)
```

Comment:
Instead of using `auto()` for the enumerations values, you can directly use the index of the direction:
```py
class Direction3D(Enum):
    """Axis directions in the 3D spacetime diagram."""

    X = 0
    Y = 1
    Z = 2
    # ...
```
in which case this method becomes 
```suggestion
    @property
    def axis_index(self) -> int:
        """Get the axis index."""
        return self.value
```
Note that, if I understood correctly, `auto` is already doing that but starts by `1` instead of `0`, so you should be able to simply do
```suggestion
    @property
    def axis_index(self) -> int:
        """Get the axis index."""
        return self.value - 1
```
here. I find that quite confusing, so I would advocate against it.